### PR TITLE
gatekeeper: rate limit log entries

### LINKED ISSUE #19: Limit rate of log entries
Gatekeeper writes log entries to report general information, warnings, and diagnosis information. These log entries are essential for operating a Gatekeeper deployment. However, during high loads or DDoS attacks, there is a potential for the log subsystem to become too heavy of a burden and impact the capacity of Gatekeeper servers. Especially, if attackers find an easy way to trigger log entries. To dodge this risk entirely, Gatekeeper must rate limit the log entries.

The Linux kernel has already a solution for this issue, namely functions: net_err_ratelimited(), net_warn_ratelimited(), net_notice_ratelimited(), among others. The proposed solution here is to study these functions to verify if we can/should adopt them in Gatekeeper. If not, we'll need to devise our own solution for this problem.

### PR DESCRIPTION
**Pull request by mengxiang0811**
This pull request closes #19

### CODE REVIEW

**Comment by AltraMayor** [comment]
If merged, doesn't this pull request close issue #19? If so, update the description of the patch.

File: Makefile [hunk 1]
```diff
@@ -48,7 +48,8 @@ SRCS-y += lib/mailbox.c lib/net.c lib/flow.c lib/ipip.c \
 
 LDLIBS += $(LDIR) -Bstatic -lluajit-5.1 -Bdynamic -lm -lmnl
 CFLAGS += $(WERROR_FLAGS) -I${GATEKEEPER}/include -I/usr/local/include/luajit-2.0/
-EXTRA_CFLAGS += -O3 -g -Wfatal-errors -DALLOW_EXPERIMENTAL_API
+EXTRA_CFLAGS += -O3 -g -Wfatal-errors -DALLOW_EXPERIMENTAL_API \
+	-DLOG_RATELIMIT_INTERVAL=5 -DLOG_RATELIMIT_BURST=10
```
Comment (AltraMayor) [inline Makefile]: 1. These parameters should come from each block configuration like the level is currently set.

2. Blocks can now change their log level dynamically. Since the level is part of the data structure of the log rate, we need to change the Lua function that updates the log level to update both.

3. What are the values of these parameters that the kernel uses? How did you derive Gatekeeper values?

File: include/gatekeeper_main.h [hunk 2]
```diff
@@ -41,4 +43,16 @@ extern uint64_t cycles_per_sec;
 extern uint64_t cycles_per_ms;
 extern uint64_t picosec_per_cycle;
 
+struct log_ratelimit_state *get_log_ratelimit_state_by_lcore(void);
+
+#define RTE_LOG_RATELIMIT(l, t, ...)                        \
```
Comment (AltraMayor) [inline include/gatekeeper_main.h]: 1. This macro seems big. Could/should not it become a function? If so, `get_log_ratelimit_state_by_lcore()` doesn't need to be public.

2. The blocks now have macros for their own logging, so you need to update those macros to use the rate limit.

File: lib/ratelimit.c [hunk 3]
```diff
@@ -63,3 +63,50 @@ tb_ratelimit_allow(struct token_bucket_ratelimit_state *tbrs)
...
+
+	if (!lrs->interval)
+		return true;
+
+	now = rte_rdtsc();
+
+	if (!lrs->begin)
+		lrs->begin = now;
+
+	if (lrs->begin + lrs->interval * cycles_per_ms < now) {
```
Comment (AltraMayor) [inline lib/ratelimit.c]: Why don't we rename `interval` to `interval_cycles` and initialize it with `lrs->interval * cycles_per_ms` to avoid this multiplication every call since `interval` is only really used here?

File: lib/ratelimit.c [hunk 4]
```diff
@@ -63,3 +63,50 @@ tb_ratelimit_allow(struct token_bucket_ratelimit_state *tbrs)
...
+{
+	uint64_t now;
+
+	if (!lrs->interval)
+		return true;
+
+	now = rte_rdtsc();
+
+	if (!lrs->begin)
+		lrs->begin = now;
```
Comment (AltraMayor) [inline lib/ratelimit.c]: Shouldn't we initialize `begin` with `rte_rdtsc()` in `log_ratelimit_state_init()` and get rid of this whole `if` satement?

File: include/gatekeeper_ratelimit.h [hunk 5]
```diff
@@ -33,8 +33,32 @@ struct token_bucket_ratelimit_state {
 	uint64_t stamp;
 };
 
+struct log_ratelimit_state {
+	uint32_t ratelimit_level;
+	uint32_t interval;
+	uint32_t burst;
+	uint32_t printed;
+	uint32_t missed;
```
Comment (AltraMayor) [inline include/gatekeeper_ratelimit.h]: The field `missed` is never used.

File: lib/ratelimit.c [hunk 6]
```diff
@@ -63,3 +63,50 @@ tb_ratelimit_allow(struct token_bucket_ratelimit_state *tbrs)
...
+
+	if (lrs->begin + lrs->interval * cycles_per_ms < now) {
+		lrs->begin = now;
+		lrs->printed = 0;
+		lrs->missed = 0;
+	}
+
+	if (lrs->burst && lrs->burst > lrs->printed) {
```
Comment (AltraMayor) [inline lib/ratelimit.c]: We can drop the first part of the test expression: `lrs->burst &&`. The reason for this is that, if `burst` is not zero, the first part of the expression doesn't make a difference. If `burst` is zero, `lrs->burst > lrs->printed` too is always false because field `printed` is of type `uint32_t` and zero cannot be bigger than any other number `printed` can have.

File: include/gatekeeper_ratelimit.h [hunk 7]
```diff
@@ -33,8 +33,32 @@ struct token_bucket_ratelimit_state {
 	uint64_t stamp;
 };
 
+struct log_ratelimit_state {
+	uint32_t ratelimit_level;
```
Comment (AltraMayor) [inline include/gatekeeper_ratelimit.h]: `struct log_ratelimit_state` must be cache aligned to avoid contention between cores.

File: lib/ratelimit.c [hunk 8]
```diff
@@ -63,3 +63,50 @@ tb_ratelimit_allow(struct token_bucket_ratelimit_state *tbrs)
...
+ * Returns:
+ * - true means go ahead and do it.
+ * - false means callbacks will be suppressed.
+ */
+bool
+log_ratelimit_allow(struct log_ratelimit_state *lrs)
+{
+	uint64_t now;
+
+	if (!lrs->interval)
```
Comment (AltraMayor) [inline lib/ratelimit.c]: DPDK Style: `lrs->interval == 0`.

File: main/main.c [hunk 9]
```diff
@@ -139,13 +141,33 @@ run_signal_handler(void)
...
 
+struct log_ratelimit_state *
+get_log_ratelimit_state_by_lcore(void)
+{
+	return &log_ratelimit_states[rte_lcore_id()];
+}
+
+static void
+log_ratelimit_init(void)
+{
```
Comment (AltraMayor) [inline main/main.c]: We don't need `log_ratelimit_init()` because the variable `log_ratelimit_states` is global, so it's automatically initialized with zeros, and `log_ratelimit_allow()` ignores states whose field `interval` is zero. Each block will set its own rate limit. Until then, there should be no rate limit.

File: Makefile (hunk 1 above)
Comment (mengxiang0811) [reply Makefile]: 3. The values are from the kernel [net_ratelimit()](https://github.com/AltraMayor/XIA-for-Linux/blob/337f3cfb501271ca3751f9ac690d2dd1951342d9/net/core/utils.c#L36), which is an overall limit. However, for gatekeeper, we will make each functional block configurable in terms of these values. As each functional block may require different logs, should we tag this with ``XXX`` for setting each block individually?

File: include/gatekeeper_main.h (hunk 2 above)
Comment (mengxiang0811) [reply include/gatekeeper_main.h]: Will do.

**Comment by mengxiang0811** [comment]
This rebase incorporate the above suggestions, and applied to CPS block. Once it's ready, I will apply the ratelimit to all other blocks.

File: main/main.c [hunk 10]
```diff
@@ -105,6 +107,25 @@ rte_strdup(const char *type, const char *s)
...
+	va_list ap;
+	int ret = 0;
+	struct log_ratelimit_state *lrs =
+		&log_ratelimit_states[rte_lcore_id()];
+
+	va_start(ap, format);
+
+	if (level < lrs->ratelimit_level)
+		ret = rte_vlog(level, logtype, format, ap);
+	else if (log_ratelimit_allow(lrs))
```
Comment (AltraMayor) [inline main/main.c]: I'm confused by these `if` statements. Can you explain the behavior you intend to have?

File: main/main.c [hunk 11]
```diff
@@ -49,6 +49,8 @@ uint64_t cycles_per_sec;
 uint64_t cycles_per_ms;
 uint64_t picosec_per_cycle;
 
+struct log_ratelimit_state log_ratelimit_states[RTE_MAX_LCORE];
+
```
Comment (AltraMayor) [inline main/main.c]: Move `log_ratelimit_states[]` to just before `rte_log_ratelimit()`, and make it `satic`.

File: include/gatekeeper_main.h [hunk 12]
```diff
@@ -29,6 +29,8 @@
 #define DEFAULT_HASH_FUNC       rte_jhash
 #endif
 
+#include "gatekeeper_ratelimit.h"
+
```
Comment (AltraMayor) [inline include/gatekeeper_main.h]: Why do we need to include the header `"gatekeeper_ratelimit.h"` here?

File: include/gatekeeper_cps.h [hunk 13]
```diff
@@ -28,8 +28,9 @@
 
 extern int cps_logtype;
 
-#define CPS_LOG(level, ...) \
-	rte_log(RTE_LOG_ ## level, cps_logtype, "GATEKEEPER CPS: " __VA_ARGS__)
+#define CPS_LOG_RATELIMIT(level, ...)                    \
+	rte_log_ratelimit(RTE_LOG_ ## level, cps_logtype,\
```
Comment (AltraMayor) [inline include/gatekeeper_cps.h]: Why did you rename `CPS_LOG()` to `CPS_LOG_RATELIMIT()` instead of keeping the name? Wouldn't replacing `rte_log()` with `rte_log_ratelimit()` be enough?

File: include/gatekeeper_main.h [hunk 14]
```diff
@@ -47,6 +47,8 @@ extern uint64_t cycles_per_sec;
 extern uint64_t cycles_per_ms;
 extern uint64_t picosec_per_cycle;
 
+extern struct log_ratelimit_state log_ratelimit_states[RTE_MAX_LCORE];
```
Comment (AltraMayor) [inline include/gatekeeper_main.h]: This prototype is suggesting that `log_ratelimit_states` should go to the ratelimit library. In fact, why don't we have a library for log ratelimit since it has no relationship with the current ratelimit library?

File: lib/ratelimit.c [hunk 15]
```diff
@@ -104,3 +104,14 @@ log_ratelimit_allow(struct log_ratelimit_state *lrs)
...
+
+void
+update_log_ratelimit_level(uint32_t level, unsigned int lcore_id)
+{
+	struct log_ratelimit_state *lrs;
+
+	RTE_VERIFY(lcore_id < RTE_MAX_LCORE);
+
+	lrs = &log_ratelimit_states[lcore_id];
+	lrs->ratelimit_level = level;
```
Comment (AltraMayor) [inline lib/ratelimit.c]: Shouldn't we update the loglevel information in DPDK as well here to avoid a mismatch?

File: Makefile (hunk 1 above)
Comment (AltraMayor) [reply Makefile]: The kernel uses `5 * HZ` instead of `5`, doesn't that mean 5s instead of 5ms?

File: main/main.c (hunk 10 above)
Comment (mengxiang0811) [reply main/main.c]: The first ``if`` statement means that the log level specified  has higher priority than the ratelimit log level, so we won't put any ratelimit on these logs.

The second ``if`` statement means that even we put ratelimit on the current log, however, the ratelimit algorithm **allows** to output a log entry, then we go ahead and call a ``rte_vlog()`` to log.

File: include/gatekeeper_cps.h (hunk 13 above)
Comment (mengxiang0811) [reply include/gatekeeper_cps.h]: As we are ratelimiting the log entries, I added ``_RATELIMIT`` to reflect this fact. I am OK with keeping the name.

File: include/gatekeeper_main.h (hunk 14 above)
Comment (mengxiang0811) [reply include/gatekeeper_main.h]: Good point!

File: Makefile (hunk 1 above)
Comment (mengxiang0811) [reply Makefile]: Good catch! In the kernel, it uses [jiffies](https://github.com/AltraMayor/XIA-for-Linux/blob/337f3cfb501271ca3751f9ac690d2dd1951342d9/lib/ratelimit.c#L46) to represent the intervals, which is defined using [HZ](https://github.com/AltraMayor/XIA-for-Linux/blob/337f3cfb501271ca3751f9ac690d2dd1951342d9/include/linux/jiffies.h#L167).

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: main/main.c (hunk 10 above)
Comment (AltraMayor) [reply main/main.c]: That means that, if Gatekeeper starts to massively generate high priority log entries, the rate limit won't help to keep the log under a reasonable rate. My suggestion is to not log if `level > lrs->ratelimit_level` as DPDK already does, and to test `log_ratelimit_allow(lrs)` otherwise.

File: lib/log_ratelimit.c [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+/*
+ * Rate limiting log entries.
+ *
+ * Returns:
+ * - true means go ahead and do it.
+ * - false means callbacks will be suppressed.
+ */
+bool
+log_ratelimit_allow(struct log_ratelimit_state *lrs)
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: `log_ratelimit_allow()` is only used in `rte_log_ratelimit()`, shouldn't it be `static`?

File: include/gatekeeper_log_ratelimit.h [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+ *   macro, so it cannot be an integer value.
+ * @param ...
+ *   The fmt string, as in printf(3), followed by the variable arguments
+ *   required by the format.
+ * @return
+ *   - 0: Success.
+ *   - Negative on error.
+ */
+#define RTE_LOG_RATELIMIT(l, t, ...)                     \
+	rte_log_ratelimit(RTE_LOG_ ## l,                 \
```
Comment (AltraMayor) [inline include/gatekeeper_log_ratelimit.h]: Macro `RTE_LOG_RATELIMIT()` is never used. Moreover, we don't want to use it in Gatekeeper since each block has its own macro. Let's drop it.

File: include/gatekeeper_log_ratelimit.h [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+#include <rte_memory.h>
+
+/*
+ * The code of this file is mostly a copy of the Linux kernel.
+ */
+
+struct log_ratelimit_state {
+	uint32_t ratelimit_level;
+	uint64_t interval_cycles;
```
Comment (AltraMayor) [inline include/gatekeeper_log_ratelimit.h]: Once `log_ratelimit_allow()` become `static` (see another comment), we can move `struct log_ratelimit_state` to `lib/log_ratelimit.c`.

File: include/gatekeeper_log_ratelimit.h [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LOG_RATELIMIT_H_
+#define _GATEKEEPER_LOG_RATELIMIT_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <rte_memory.h>
```
Comment (AltraMayor) [inline include/gatekeeper_log_ratelimit.h]: Once other comments are addressed, we should be able to move `#include <stdbool.h>` and `#include <rte_memory.h>` to `lib/log_ratelimit.c`.

File: main/main.c (hunk 10 above)
Comment (AltraMayor) [reply main/main.c]: Moreover, if we call `rte_log_get_level()` to get the log level, we can discard field `lrs->ratelimit_level` and avoid the trouble of synchronizing both level variables: `lrs->ratelimit_level` and its DPDK equivalent.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: lua/gatekeeper.lua [hunk 20]
```diff
@@ -271,6 +273,9 @@ uint32_t rte_log_get_global_level(void);
 int rte_log_set_level(uint32_t type, uint32_t level);
 int rte_log_get_level(uint32_t type);
 
+int update_log_ratelimit_level(uint32_t type,
+	uint32_t level, unsigned int lcore_id);
+
```
Comment (AltraMayor) [inline lua/gatekeeper.lua]: This prototype is never defined or used. Drop it.

File: lua/cps.lua [hunk 21]
```diff
@@ -28,6 +28,11 @@ return function (net_conf, gk_conf, gt_conf, lls_conf, numa_table)
 	-- to output routing table updates to the log.
 	cps_conf.log_level = gatekeeper.c.RTE_LOG_DEBUG
 
+	-- XXX #155 Sample parameters, need to be tested for better performance.
+	-- Log ratelimit interval and burst size.
+	cps_conf.log_ratelimit_interval_ms = 5000
```
Comment (AltraMayor) [inline lua/cps.lua]: Drop the XXX tag. These parameters are okay and have little impact on the general performance of Gatekeeper. If someone ever wants to change them, it'll likely be in response to convenience.

This comment affects other patches in this pull request.

**Comment by AltraMayor** [comment]
Squash patches **ratelimit: add functionalities to limit the rate of log entries** and **cps: limit rate of log entries** together to avoid breaking the binary search of Git. The combined patch should be called **cps: limit rate of log entries**.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: lib/log_ratelimit.c [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+		return true;
+	}
+
+	return false;
+}
+
+int
+rte_log_ratelimit(uint32_t level, uint32_t logtype, const char *format, ...)
+{
+	va_list ap;
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Move `ap` into the `if` statement in which is used.

File: lib/log_ratelimit.c [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+	struct log_ratelimit_state *lrs;
+	int ratelimit_level = rte_log_get_level(logtype);
+	if (ratelimit_level < 0)
+		return -1;
+
+	lrs = &log_ratelimit_states[rte_lcore_id()];
+	if (level <= (uint32_t)ratelimit_level && log_ratelimit_allow(lrs)) {
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Replace `uint32_t` with `typeof(level)`.

File: lib/log_ratelimit.c [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+		int ret;
+		va_start(ap, format);
+
+		ret = rte_vlog(level, logtype, format, ap);
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Style: 1. add a blank line before `va_start()`, 2. remove blank lines before and after `rte_vlog()`.

File: lib/log_ratelimit.c [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+{
+	struct log_ratelimit_state *lrs;
+
+	RTE_VERIFY(lcore_id < RTE_MAX_LCORE);
+
+	lrs = &log_ratelimit_states[lcore_id];
+	lrs->interval_cycles = interval * cycles_per_ms;
+	lrs->burst = burst;
+	lrs->printed = 0;
+	lrs->begin = rte_rdtsc();
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Replace the two last lines above with `log_ratelimit_reset(lrs, rte_rdtsc());`.

See another comment for details.

File: lib/log_ratelimit.c [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+	uint64_t now;
+
+	if (lrs->interval_cycles == 0)
+		return true;
+
+	now = rte_rdtsc();
+
+	if (lrs->begin + lrs->interval_cycles < now) {
+		lrs->begin = now;
+		lrs->printed = 0;
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Let's rename `begin` with `end`, and do the following:
1. the test above becomes `lrs->end < now`; and
2. replace the body of this `if` statement with `log_ratelimit_reset(lrs, now);`. We will no longer need the braces for the body of this `if` statement;
3. Define `log_ratelimit_reset()` before `log_ratelimit_state_init()` as follows:
```
static inline void
log_ratelimit_reset(struct log_ratelimit_state *lrs, uint64_t now)
{
	lrs->printed = 0;
	lrs->end = now + lrs->interval_cycles;
}
```

File: include/gatekeeper_log_ratelimit.h [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+ /*
+  * @lcore_id: initialize the log_ratelimit_state data for @lcore_id.
+  *
+  * This will allow to enforce a rate limit on log entries:
+  * no more than @log_ratelimit_state.burst callbacks in
+  * every @log_ratelimit_state.interval milliseconds.
+  *
+  * Note that, to avoid performance degradation caused by locks, the
+  * implementation assumes that each lcore will maintain a separate
+  * log_ratelimit_state data to rate limit the log entries.
```
Comment (AltraMayor) [inline include/gatekeeper_log_ratelimit.h]: Replace `log_ratelimit_state data` with `struct log_ratelimit_state`.

File: lib/log_ratelimit.c [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+	return false;
+}
+
+int
+rte_log_ratelimit(uint32_t level, uint32_t logtype, const char *format, ...)
+{
+	va_list ap;
+	struct log_ratelimit_state *lrs;
+	int ratelimit_level = rte_log_get_level(logtype);
+	if (ratelimit_level < 0)
```
Comment (AltraMayor) [inline lib/log_ratelimit.c]: Add `ublikely()` to the test.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.
